### PR TITLE
feat: Add integration test for mobile HTTP calls

### PR DESCRIPTION
**Pull request by jrconlin**
This adds a "stub" router, specifically for integration testing. (May
put this behind a feature flag later.) Depending on the "client" type
the stub router will either succeed or fail.

This also adds some tests to ensure that mobile devices get a proper
endpoint depending on presence of a `key` value.

Closes SYNC-4041

### CODE REVIEW

File: autoendpoint/src/routers/stub/router.rs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+            .router_data
+            .as_ref()
+            .ok_or(StubError::General("No Registration".to_owned()))?;
+
+        let default = Value::String("success".to_owned());
+        let client_type = router_data.get("type").unwrap_or(&default);
+        let client = self
+            .clients
+            .get(&client_type.to_string().to_lowercase())
+            .unwrap_or_else(|| self.clients.get("success").unwrap());
```
Comment (pjenvey) [inline autoendpoint/src/routers/stub/router.rs:81]: Maybe default to erroring instead?
```suggestion
            .unwrap_or_else(|| self.clients.get("error").unwrap());
```

File: autoendpoint/src/routers/stub/settings.rs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    #[serde(rename = "credentials")]
+    pub server_credentials: String,
+}
+
+/// Credential information for each application
+#[derive(Clone, Debug, serde::Deserialize, serde::Serialize)]
+pub struct StubServerSettings {
+    pub project_id: String,
+    pub error: String,
+}
```
Comment (pjenvey) [inline autoendpoint/src/routers/stub/settings.rs:16]: The creds aren't used at all, kill?
```suggestion
}

```

**Review by pjenvey — CHANGES_REQUESTED** [review]
I think this needs the option of disabling it completely on production, otherwise we could be allowing "stub" `router_type`s to exist, particularly in the db, which isn't ideal.

This is a little annoying to accomplish due to routers pretty much being "static" along with their `RouterType` enum but I think I accomplished it on [this branch](https://github.com/mozilla-services/autopush-rs/compare/feat/sync-4041-http...feat/sync-4041-http-disable), as there only seems to be the one extractor that actually parses/validates them.

File: autoendpoint/src/routers/stub/settings.rs (hunk 2 above)
Comment (jrconlin) [reply autoendpoint/src/routers/stub/settings.rs:16]: Ah! Thanks, forgot to make the change to router.

**Comment by jrconlin** [comment]
I agree that this should probably have some flag that prevents it from going to production. I've added a `stub` feature flag.

File: autoendpoint/src/routers/stub/router.rs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                StubError::General(format!("Unknown App ID: {}", app_id.to_owned())).into(),
+            );
+        }
+
+        let mut router_data = HashMap::new();
+        router_data.insert(
+            "token".to_string(),
+            serde_json::to_value(&router_data_input.token).unwrap(),
+        );
+        router_data.insert("app_id".to_string(), serde_json::to_value(app_id).unwrap());
```
Comment (pjenvey) [inline autoendpoint/src/routers/stub/router.rs:65]: A minor nit, but there's a mix of `to_string` and `to_owned` in this PR, let's just stick with one
```suggestion
            "token".to_owned(),
            serde_json::to_value(&router_data_input.token).unwrap(),
        );
        router_data.insert("app_id".to_owned(), serde_json::to_value(app_id).unwrap());
```

File: autoendpoint/src/routers/stub/settings.rs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+            error: "General Error".to_owned(),
+        }
+    }
+}
+
+impl Default for StubSettings {
+    fn default() -> Self {
+        Self {
+            url: "http://localhost:8080".to_owned(),
+            server_credentials: "{\"error\":\"General Error\"}".to_string(),
```
Comment (pjenvey) [inline autoendpoint/src/routers/stub/settings.rs:30]:
```suggestion
            url: "http://localhost:8080".to_owned(),
            server_credentials: "{\"error\":\"General Error\"}".to_owned(),
```

**Review by pjenvey — CHANGES_REQUESTED** [review]

**Review by pjenvey — DISMISSED** [review]
Just needs the conflicts resolved

**Review by pjenvey — APPROVED** [review]

**Pull request merged**